scripts/configuration_helper.lua

```diff
@@ -86,820 +72,386 @@ function sgct.config.single(arg) end
-            "capture['format'] must be a string or nil"
-        )
-    end
-
-    assert(type(arg["viewport"]) == "string", "viewport must be a string")
-
-    -- Then setting reasonable default values
-    if arg["vsync"] == nil then
-        arg["vsync"] = false
-    end
-
-    if arg["fullScreen"] == nil then
-        arg["fullScreen"] = false
-    end
-
-    if arg["monitor"] == nil then
-        arg["monitor"] = 0
-    end
-    
-    if arg["tags"] == nil then
-        arg["tags"] = {}
-    end
-
-    if arg["msaa"] == nil then
-        arg["msaa"] = 4
-    end
-
-    if arg["border"] == nil then
-        arg["border"] = true
-    end
-
-     if arg["shared"] then
-        local t = arg["tags"]
-        t[#t + 1] = "Spout"
-    end
-
-    if arg["name"] == nil then
-        arg["name"] = "OpenSpace"
-    end
-
-    if arg["stereo"] == nil then
-        arg["stereo"] = "none"
-    end
-
-    if arg["windowPos"] == nil then
-        arg["windowPos"] = { 50, 50 }
-    end
-
-    if arg["eyeSep"] == nil then
-        arg["eyeSep"] = 0.065
-    end
+function generateCluster(arg, viewport)
+  local result = {}
+
+  table.insert(result,   [[{]])
+  table.insert(result,   [[  "version": 1,]])
+  table.insert(result,   [[  "masterAddress": "127.0.0.1",]])
+
+  if arg["sgctDebug"] then
+    table.insert(result, [[  "debug": ]] .. tostring(arg["sgctDebug"]) .. ",")
+  end
+
+  if (arg["settings"]) then
+    table.insert(result, [[  "settings": {]])
+    generateSettings(result, arg["refreshRate"], arg["vsync"])
+    table.insert(result, [[  },]])
+  end
+
+  if arg["scene"] then
+    table.insert(result, [[  "scene": {]])
+    generateScene(result, arg["scene"])
+    table.insert(result, [[  },]])
+  end
+
+  table.insert(result,   [[  "nodes": []])
+  table.insert(result,   [[    {]])
+  table.insert(result,   [[      "address": "127.0.0.1",]])
+  table.insert(result,   [[      "port": 20401,]])
+  table.insert(result,   [[      "windows": []])
+  table.insert(result,   [[        {]])
+  generateWindow(result, arg["fullScreen"], arg["msaa"], arg["border"], arg["monitor"],
+    arg["tags"], arg["stereo"], arg["pos"], arg["size"], arg["res"], viewport)
+  table.insert(result,   [[        }]])
+  table.insert(result,   [[      ] ]])
+  table.insert(result,   [[    }]])
+  table.insert(result,   [[  ],]])
+  table.insert(result,   [[  "users": []])
+  generateUser(result)
+  table.insert(result,   [[  ] ]])
+  table.insert(result,   [[}]])
+
+  return table.concat(result, "\n")
+end
 
-    if arg["eyePos"] == nil then
-        arg["eyePos"] = { 0.0, 0.0, 0.0 }
-    end
 
-    if arg["sgctDebug"] == nil then
-        arg["sgctDebug"] = false
-    end
 
-    if not arg["windowSize"] then
-        arg["windowSize"] = {1280, 720}
-    end
+function check(type_str, arg, param, subparam_type)
+  local t = type(arg[param])
+  assert(t == type_str or t == "nil", param .. " must be a " .. type_str .. " or nil")
```

Comment:
I might be wrong but shouldn't the `t == type_str` actually be `t ~= type_str`? To catch when it is **not** the type you want?

Reply:
I don't think so?   We are asserting that the type is correct or nil?